ok, getting closer but:

```
../Nim/bin/nim: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by ../Nim/bin/nim)
Command exited with non-zero status 1
0.00user 0.00system 0:00.00elapsed 33%CPU (0avgtext+0avgdata 604maxresident)k
6192inputs+0outputs (1major+49minor)pagefaults 0swaps
make: *** [nlvm/nlvmr] Error 1
```

I noticed that the build environment, including nim and llvm, is being copied into the docker - this is likely where the problem lies: binaries from the docker host should not be used inside the docker so they need to be excluded when copying stuff to docker!

to repro, build nlvm using `make`, then run `make docker` - if lots and lots of GB of data are being transferred into the docker, something is wrong.

--- a/.dockerignore
+++ b/.dockerignore
@@ -1,0 +1,7 @@
+ext/*
+nlvm/nlvm
+nlvm/nlvmr
+nlvm/nlvm.ll
+nlvm/nlvm.self.ll
+nlvm/nlvm.self
+Nim

--- a/Dockerfile
+++ b/Dockerfile
@@ -1,0 +1,23 @@
+FROM ubuntu:18.04 AS build
+
+ADD . /code
+WORKDIR /code
+
+RUN apt-get update && \
+    apt-get install -y rsync time git apt-utils libpcre3-dev libtinfo-dev libz-dev libssl-dev libsqlite3-dev build-essential cmake ninja-build python-minimal wget && \
+    rm -rf /var/lib/apt/lists/* && \
+    git submodule update --init && \
+    ./make-dist-docker.sh && \
+    cd / && \
+    rm -rf /code/ && \
+    apt-get remove -y git apt-utils libpcre3-dev libssl-dev libsqlite3-dev build-essential cmake ninja-build python-minimal wget
+
+# Create final image
+FROM ubuntu:18.04
+
+COPY --from=build /usr/bin/nlvm /usr/bin/nlvm
+COPY --from=build /usr/lib/ /usr/lib/
+RUN apt-get update && apt-get install -y gcc && rm -rf /var/lib/apt/lists/*
+
+ENTRYPOINT ["/usr/bin/nlvm"]
+

--- a/Makefile
+++ b/Makefile
@@ -106,3 +106,7 @@ ext/$(LLVM_DIR)/sha/lib/libLLVM-$(LLVM_MAJ).so:
 
 ext/$(LLVM_DIR)/sta/bin/llvm-config:
 	sh ./dl-llvm.sh $(LLVM_MAJ) $(LLVM_MIN) $(LLVM_PAT) sta
+
+.PHONY: docker
+docker:
+	docker build . -t nlvm --no-cache

--- a/README.md
+++ b/README.md
@@ -107,6 +107,15 @@ When you update `nlvm` from `git`, don't forget the submodule:
 
     git pull && git submodule update
 
+To build a docker image, use:
+  
+    make docker
+
+To run built `nlvm` docker image use:
+    
+    docker run -v $(pwd):/code/ nlvm c -r /code/test.nim
+
+
 # Compiling your code
 
 On the command line, `nlvm` is mostly compatible with `nim`.

--- a/make-dist-docker.sh
+++ b/make-dist-docker.sh
@@ -1,0 +1,27 @@
+#!/bin/bash
+
+# Experimental release script
+
+set -e
+
+ROOT=nlvm-build-root
+
+rm -rf $ROOT
+
+# Make sure the nlvm binary is fresh
+rm -f nlvm/nlvmr
+make STATIC_LLVM=1 nlvm/nlvmr
+
+# Copy nlvm and library files
+# TODO these would go in /usr/{bin, share/Nim} normally
+mkdir -p $ROOT
+cp nlvm/nlvmr $ROOT/nlvm
+cp -r nlvm-lib $ROOT
+mkdir -p $ROOT/Nim
+cd Nim
+# avoid build junk
+git archive --format=tar HEAD lib config compiler doc | (cd ../$ROOT/Nim && tar xf -)
+cd ..
+mkdir -p /usr/lib/nlvm
+cp -av $ROOT/* /usr/lib/nlvm
+ln -s /usr/lib/nlvm/nlvm /usr/bin/nlvm

